src/Polly.Core/Strategy/Outcome.TResult.cs

```diff
@@ -0,0 +1,88 @@
+using System.Diagnostics.CodeAnalysis;
+
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+
+namespace Polly.Strategy;
+
+/// <summary>
+/// Represents the outcome of an operation that returns a result of type TResult or an exception.
+/// </summary>
+/// <typeparam name="TResult">The type of the result produced by the operation.</typeparam>
+public readonly struct Outcome<TResult>
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Outcome{TResult}"/> struct with the specified exception.
+    /// </summary>
+    /// <param name="exception">The exception that occurred during the operation.</param>
+    public Outcome(Exception exception)
+        : this() => Exception = Guard.NotNull(exception);
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Outcome{TResult}"/> struct with the specified result.
+    /// </summary>
+    /// <param name="result">The result produced by the operation.</param>
+    public Outcome(TResult result)
+        : this() => Result = result;
+
+    /// <summary>
+    /// Gets the exception that occurred during the operation, if any.
+    /// </summary>
+    public Exception? Exception { get; }
+
+    /// <summary>
+    /// Gets the result produced by the operation, if any.
+    /// </summary>
+    public TResult? Result { get; }
```

Comment:
Can we add "not null when" attributes to these that look to `HasResult`?

Reply:
I think that still would be misleading. We can still do:

``` csharp
new Outcome<MyClass>(null!)
```

`null` for some scenarios can be considered a valid result. So `HasResult` is true but the result is null.

Which actually means this signature is invalid:

``` csharp
    public bool TryGetResult([NotNullWhen(true)] out TResult? result)
```
I'll drop `NotNullWhen`.